Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp

```diff
@@ -140,13 +165,113 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveReplaceme
+
+    if (!m_webPage) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    auto* corePage = m_webPage->corePage();
+    if (!corePage) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    Ref frame = CheckedRef(corePage->focusController())->focusedOrMainFrame();
+    RefPtr document = frame->document();
+    if (!document) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    if (state != WebTextReplacementData::State::Committed && state != WebTextReplacementData::State::Reverted)
+        return;
+
+    auto nodeAndMarker = findReplacementMarkerByUUID(document.get(), replacement.uuid);
+    if (!nodeAndMarker) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    auto& [node, marker] = *nodeAndMarker;
+
+    auto rangeToReplace = makeSimpleRange(node, marker);
+
+    auto oldData = std::get<DocumentMarker::UnifiedTextReplacementData>(marker.data());
+
+    auto offsetRange = OffsetRange { marker.startOffset(), marker.endOffset() };
+    document->markers().removeMarkers(node, offsetRange, { DocumentMarker::Type::UnifiedTextReplacement });
+
+    auto [newText, newState] = [&]() -> std::tuple<String, DocumentMarker::UnifiedTextReplacementData::State> {
+        switch (state) {
+        case WebTextReplacementData::State::Committed:
+            return std::make_tuple(replacement.replacement, DocumentMarker::UnifiedTextReplacementData::State::Committed);
+
+        case WebTextReplacementData::State::Reverted:
+            return std::make_tuple(oldData.originalText, DocumentMarker::UnifiedTextReplacementData::State::Reverted);
+
+        default:
+            RELEASE_ASSERT_NOT_REACHED();
+        }
+    }();
+
+    replaceTextInRange(frame.get(), rangeToReplace, newText);
+
+    auto newData = DocumentMarker::UnifiedTextReplacementData { oldData.originalText, oldData.uuid, newState };
+    document->markers().addMarker(node, DocumentMarker { DocumentMarker::Type::UnifiedTextReplacement, offsetRange, WTFMove(newData) });
+}
+
+void UnifiedTextReplacementController::didEndTextReplacementSession(const WTF::UUID& uuid, bool accepted)
+{
+    RELEASE_LOG(UnifiedTextReplacement, "UnifiedTextReplacementController::didEndTextReplacementSession (%s) [accepted: %d]", uuid.toString().utf8().data(), accepted);
+
+    if (!m_webPage) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    auto* corePage = m_webPage->corePage();
+    if (!corePage) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    Ref frame = CheckedRef(corePage->focusController())->focusedOrMainFrame();
+    RefPtr document = frame->document();
+    if (!document) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    // FIXME: Associate the markers with a specific session, and only modify the markers which belong
+    // to this session.
```

Comment:
This is an optimization, and not a correctness `FIXME` right?

In any case, a filing a bug would be good.